SpatialGDK/Source/SpatialGDK/Private/Interop/ClientNetLoadActorHelper.cpp

```diff
@@ -0,0 +1,142 @@
+			if (SubObjectNetGUID.IsValid())
+			{
+				FUnrealObjectRef SubObjectRef = NetDriver->PackageMap->GetUnrealObjectRefFromNetGUID(SubObjectNetGUID);
+				if (SubObjectRef.IsValid() && IsDynamicSubObject(*NetDriver, Actor, SubObjectRef.Offset))
+				{
+					SaveDynamicSubObjectRef(SubObjectRef, SubObjectNetGUID);
+					UE_LOG(
+						LogClientNetLoadActorHelper, Verbose,
+						TEXT("Saved reusable dynamic SubObject ObjectRef (ObjectRef offset: %u) for ClientNetLoad actor with entityId %d"),
+						SubObjectRef.Offset, SubObjectRef.Entity);
+				}
+			}
+		}
+	}
+}
+
+FNetworkGUID* FClientNetLoadActorHelper::GetSavedDynamicSubObjectNetGUID(const FUnrealObjectRef& ObjectRef)
+{
+	if (TMap<FUnrealObjectRef, FNetworkGUID>* EntityMappings = DynamicSubObjectRefToGuid.Find(ObjectRef.Entity))
+	{
+		if (FNetworkGUID* NetGUID = EntityMappings->Find(ObjectRef))
+		{
+			return NetGUID;
+		}
+	}
+	return nullptr;
+}
+
+void FClientNetLoadActorHelper::SaveDynamicSubObjectRef(const FUnrealObjectRef& ObjectRef, const FNetworkGUID& NetGUID)
+{
+	TMap<FUnrealObjectRef, FNetworkGUID>& EntityMappings = DynamicSubObjectRefToGuid.FindOrAdd(ObjectRef.Entity);
+	EntityMappings.Emplace(ObjectRef, NetGUID);
+}
+
+void FClientNetLoadActorHelper::ClearDynamicSubObjectRefs(const Worker_EntityId InEntityId)
+{
+	if (TMap<FUnrealObjectRef, FNetworkGUID>* EntityMappings = DynamicSubObjectRefToGuid.Find(InEntityId))
+	{
+		for (auto DynamicSubObjectIterator = EntityMappings->CreateIterator(); DynamicSubObjectIterator; ++DynamicSubObjectIterator)
+		{
+			DynamicSubObjectIterator.RemoveCurrent();
+		}
+	}
+}
+
+void FClientNetLoadActorHelper::RemoveRuntimeRemovedComponents(const Worker_EntityId EntityId, const TArray<ComponentData>& NewComponents)
+{
+	auto ContainedInComponentsArr = [this, EntityId, &NewComponents](const FUnrealObjectRef& CheckComponentObjRef) {
+		for (const ComponentData& Component : NewComponents)
+		{
+			// Skip if this isn't a generated component
+			if (Component.GetComponentId() < SpatialConstants::STARTING_GENERATED_COMPONENT_ID
+				&& Component.GetComponentId() != SpatialConstants::DORMANT_COMPONENT_ID)
+			{
+				continue;
+			}
+
+			uint32 Offset = 0;
+			NetDriver->ClassInfoManager->GetOffsetByComponentId(Component.GetComponentId(), Offset);
+
+			if (FUnrealObjectRef(EntityId, Offset) == CheckComponentObjRef)
+			{
+				return true;
+			}
+		}
+		return false;
+	};
+
+	if (TMap<FUnrealObjectRef, FNetworkGUID>* EntityMappings = DynamicSubObjectRefToGuid.Find(EntityId))
+	{
+		// Go over each stored sub-object and determine whether it is contained within the new components array
+		// If it is not contained within the new components array, it means the sub-object was removed while out of the client's interest
+		// If so, remove it now
+		for (auto DynamicSubObjectIterator = EntityMappings->CreateIterator(); DynamicSubObjectIterator; ++DynamicSubObjectIterator)
+		{
+			if (DynamicSubObjectIterator->Key.Entity == EntityId)
```

Comment:
grinds my gears that we iterate map keys, but I guess it *is* rare...

Reply:
yea, i had unneeded if statement there as well so ty for highlight